Support for data-dependent output shapes

dyastremsky opened the pull request:
This pull request uses OutputAllocators to handle data-dependent output shapes.

Related server PR: https://github.com/triton-inference-server/server/pull/6057

tanmayv25 commented on src/instance_state.h:
```diff
@@ -551,6 +554,10 @@ class ModelInstanceState : public TensorRTModelInstance {
+
+    // Whether this is an output using OutputAllocator for
+    // dynamic resizing.
+    bool is_dynamic_{false};
```
is_dynamic_ -> is_dynamic_shape_output_ ?

dyastremsky replied:
Updated!

dyastremsky commented on src/output_allocator.cc:
```diff
@@ -0,0 +1,75 @@
+    // If zero copy support is enabled, need to set the buffer to the device
+    // pointer.
+    if (zero_copy_support_) {
+      void* device_buffer;
```
Good find. Done!

dyastremsky commented on src/output_allocator.cc:
```diff
@@ -0,0 +1,75 @@
+    char const* tensor_name, void* current_memory, uint64_t size,
+    uint64_t alignment) noexcept
+{
+  if (size > output_size_) {
```
Sure, added.

dyastremsky commented on src/instance_state.h:
```diff
@@ -551,6 +554,10 @@ class ModelInstanceState : public TensorRTModelInstance {
+
+    // Whether this is an output using OutputAllocator for
+    // dynamic resizing.
+    bool is_dynamic_{false};
```
Sure. I've updated the member variable initialization across the file. Guan linked [this](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-in-class-initializer) and wanted to start migrating us toward in-class initializers, though there were concerns of consistency as we started. So at least this file will use in-class initializers, and we can start updating this as we go.

tanmayv25 commented on src/instance_state.cc:
```diff
@@ -1114,20 +1122,25 @@ ModelInstanceState::Run(
             "failed to run TRT response");
       }
 
+      void* buffer =
```
Can you modify the io_binding_info struct into a class with some of the this logic exposed as function? The abstraction will also reduce code duplicity.

dyastremsky replied:
Good thinking. Done.

tanmayv25 commented on src/instance_state.h:
```diff
@@ -570,32 +520,35 @@ class ModelInstanceState : public TensorRTModelInstance {
+  std::unique_ptr<Semaphore> semaphore_{nullptr};
+
+  // Vector of OutputAllocators to hold pointers to prevent deallocation.
+  std::vector<std::unique_ptr<OutputAllocator>> allocators_{};
```
Shouldn't we move the allocators within IOBindingInfo? Each IOBindingInfo will map to a unique OutputAllocator if having dynamic output.

dyastremsky replied:
Good idea. Done.

tanmayv25 commented on src/io_binding_info.h:
```diff
@@ -0,0 +1,148 @@
+  uint64_t GetByteSize() const { return byte_size_; }
+
+  void SetBuffer(void* buffer) { buffer_ = buffer; }
+  void* GetBuffer() const { return buffer_; }
```
With OutputAllocator inside this class, the GetBuffer function can optionally use OutputAllocator or the raw buffer pointer. 
This will help us abstract away OutputAllocator logic from the instance_state class. I believe the only exposure would be SetTensorAddress or SetOutputAllocator calls.

dyastremsky replied:
Great idea. Done.

dyastremsky commented:
@tanmayv25 This ticket is now ready for another review. I have incorporated your changes.

I think we should review and merge this ticket without the related testing ticket. I tested this locally and the fix works. I am also running a CI pipeline to ensure everything passes. The reason we should do the testing ticket separately is because the new shape tensor tests I added cannot be generated for testing until the next TensorRT container is released.

While we could split that testing PR into two PRs and merge the non-shape testing pieces, the next version of TensorRT should include the fix for the unexpected behavior of getting 0 for max byte size. As a result, we can't replicate that unexpected behavior for the next TRT version and therefore there is no point to switching to using two nonzero op layers in the test anyway. In short, once the new TRT container can be used as an upstream, we can then update the tests to work with the new TRT version.

tanmayv25 commented:
I understand. Let's merge the PR and we can add testing once we move to TRT 9.0

Tabrizian commented on src/output_allocator.cc:
```diff
@@ -0,0 +1,77 @@
+namespace triton { namespace backend { namespace tensorrt {
+
+void*
+OutputAllocator::reallocateOutput(
```
```suggestion
OutputAllocator::ReallocateOutput(
```

dyastremsky replied:
We can't make this change, as this is implementing the TensorRT IOutputAllocator interface (see [here](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#dynamic-shaped-output)). I think the only function we can make match out code style is GetBuffer() and GetBufferAddr(), but then the function names within the class would be inconsistent.

Tabrizian replied:
got it, thanks for the clarification.

Tabrizian commented on src/output_allocator.h:
```diff
@@ -0,0 +1,68 @@
+  void notifyShape(
+      char const* tensor_name, nvinfer1::Dims const& dims) noexcept override;
+
+  void* getBuffer() { return output_ptr_; };
```
```suggestion
  void* GetBuffer() { return output_ptr_; };
```

Tabrizian commented on src/output_allocator.h:
```diff
@@ -0,0 +1,68 @@
+      char const* tensor_name, nvinfer1::Dims const& dims) noexcept override;
+
+  void* getBuffer() { return output_ptr_; };
+  void** getBufferAddr() { return &output_ptr_; };
```
```suggestion
  void** GetBufferAddr() { return &output_ptr_; };
```

dyastremsky commented:
Added a few quality-of-life updates. Fixed the test failure (the output allocator was getting created multiple times, so all but the last context would have had their allocator go out of scope): https://github.com/triton-inference-server/tensorrt_backend/pull/72/commits/b3705deeccc4bf2a589beb9bbbc080f4e0264cb0

L0_trt_data_dependent_shape and L0_trt_dynamic_shape passed locally. Re-running CI now.

tanmayv25 commented on src/output_allocator.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace triton { namespace backend { namespace tensorrt {
+
+class OutputAllocator : public nvinfer1::IOutputAllocator {
```
nit: Can you add a comment here about the purpose of this class and how it extends nvinfer1::IOutputAllocator?
It would avoid confusion like @Tabrizian ran into below.
